CanvasBase contains 2D context specific drawing methods

kkinnunen-apple opened the pull request:
#### ea2135ab14ff38d8adaa16c94ee1666c4b9ff2dd
<pre>
CanvasBase contains 2D context specific drawing methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=274750">https://bugs.webkit.org/show_bug.cgi?id=274750</a>
<a href="https://rdar.apple.com/128788730">rdar://128788730</a>

Reviewed by Said Abou-Hallawa.

Move drawingContext(), existingDrawingContext(), baseTransform() to
CanvasRenderingContext2DBase as they are only related to 2D context
implementation.

This is work towards CanvasRenderingContext subclasses themselves managing
the copied buffers.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::drawingContext const): Deleted.
(WebCore::CanvasBase::existingDrawingContext const): Deleted.
(WebCore::CanvasBase::baseTransform const): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::hasCreatedImageBuffer const):
* Source/WebCore/html/CustomPaintCanvas.cpp:
(WebCore::CustomPaintCanvas::getContext):
(WebCore::CustomPaintCanvas::replayDisplayList):
(WebCore::CustomPaintCanvas::copiedImage const):
(WebCore::CustomPaintCanvas::baseTransform const): Deleted.
(WebCore::CustomPaintCanvas::drawingContext const): Deleted.
(WebCore::CustomPaintCanvas::existingDrawingContext const): Deleted.
(WebCore::CustomPaintCanvas::replayDisplayListImpl const): Deleted.
* Source/WebCore/html/CustomPaintCanvas.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::didDraw):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::unwindStateStack):
(WebCore::CanvasRenderingContext2DBase::isAccelerated const):
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::clearCanvas):
(WebCore::CanvasRenderingContext2DBase::transformAreaToDevice const):
(WebCore::CanvasRenderingContext2DBase::calculateCompositingBufferRect):
(WebCore::CanvasRenderingContext2DBase::compositeBuffer):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::drawingContext const):
(WebCore::CanvasRenderingContext2DBase::existingDrawingContext const):
(WebCore::CanvasRenderingContext2DBase::baseTransform const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/PaintRenderingContext2D.cpp:
(WebCore::PaintRenderingContext2D::create):
(WebCore::PaintRenderingContext2D::PaintRenderingContext2D):
(WebCore::PaintRenderingContext2D::canvas const):
(WebCore::PaintRenderingContext2D::drawingContext const):
(WebCore::PaintRenderingContext2D::existingDrawingContext const):
(WebCore::PaintRenderingContext2D::baseTransform const):
(WebCore::PaintRenderingContext2D::replayDisplayList const):
* Source/WebCore/html/canvas/PaintRenderingContext2D.h:

Canonical link: <a href="https://commits.webkit.org/279912@main">https://commits.webkit.org/279912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3761452db96e0a31472f8bad6170d530db63ac1d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54901 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34346 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58179 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57201 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5656 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44464 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3818 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47567 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25591 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4908 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3773 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5130 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59769 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30158 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5275 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51886 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31295 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47645 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51311 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12068 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -67,36 +67,13 @@ CanvasBase::~CanvasBase()
     m_canvasNoiseHashSalt = std::nullopt;
 }
 
-GraphicsContext* CanvasBase::drawingContext() const
```
I do not understand fully the purpose of this change. If CanvasBase should not have a drawingContext, why does it still have to own an ImageBuffer? What is the use of this ImageBuffer if CanvasBase does not draw on it?

kkinnunen-apple replied:
> I do not understand fully the purpose of this change. If CanvasBase should not have a drawingContext, why does it still have to own an ImageBuffer? What is the use of this ImageBuffer if CanvasBase does not draw on it?

The end goal is exactly what you ask: for the CanvasBase to not own the image buffer, as the drawing buffer is the property of the rendering context.
I just cannot make that change without regressions, so I'm working on it piece by piece, to simplify the implementation and the review. This is mentioned in passing in the commit message.

shallawa replied:
I understood.

But please file a bug for `the CanvasBase to not own the image buffer` issue and block it by [274750](https://bugs.webkit.org/show_bug.cgi?id=274750). Also FIXME comments will help understanding your intermediate changes.

shallawa commented on Source/WebCore/html/CustomPaintCanvas.cpp:
```diff
@@ -77,27 +72,20 @@ void CustomPaintCanvas::replayDisplayList(GraphicsContext& target)
     auto& imageTarget = image->context();
     imageTarget.translate(-clipBounds.location());
-    replayDisplayListImpl(imageTarget);
+    if (m_context)
```
If m_context is null, I think we should make an early return even before creating  `image`.
```
auto image = target.createAlignedImageBuffer(...);
```
No need to draw a clear ImageBuffer to the `target`.

kkinnunen-apple replied:
No, that is a change of behavior and incorrect operation in case, for example, when drawing with certain compositing operations.

shallawa replied:
If `m_context` is nullptr then no `DisplayList` will be replayed back to the `imageTarget`. This will make `image` a transparent empty `ImageBuffer`. Why do we still need to draw this transparent empty `ImageBuffer`? What compositing operations will have an effect on drawing a transparent empty `ImageBuffer`?

Any way I think an early return here is an optimization but this should not affect the correctness.

kkinnunen-apple replied:
If the context has Copy compositing mode instead of SourceOver, the skip would mean the drawing buffer would be left uninitialized instead of initialized with transparent.
As said, the previous code was like that, and I'd like to keep it similar. The CustomPaintCanvas is buggy in various ways related to here, so if it is ever needed, it needs to be re-implemented anyway.

shallawa commented on Source/WebCore/html/CustomPaintCanvas.cpp:
```diff
@@ -77,27 +72,20 @@ void CustomPaintCanvas::replayDisplayList(GraphicsContext& target)
     auto buffer = ImageBuffer::create(size(), RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
     if (buffer) {
-        replayDisplayListImpl(buffer->context());
+        if (m_context)
+            m_context->replayDisplayList(buffer->context());
```
Ditto, early return.

kkinnunen-apple replied:
Nope, it means a different thing.
The canvas is empty when it doesn't have a context. It still produces an empty transparent image. This is important when drawing with complex compositing operations or when accessing the underlying pixels.
Also, it is not what is there in the left: we don't want change of behavior in this part of the code in this patch.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2384,14 +2388,26 @@ const Vector<CanvasRenderingContext2DBase::State, 1>& CanvasRenderingContext2DBa
+AffineTransform CanvasRenderingContext2DBase::baseTransform() const
+{
+    ASSERT(canvasBase().hasCreatedImageBuffer());
+    return canvasBase().buffer()->baseTransform();
```
This is a little bit confusing. If `CanvasBase` owns the `ImageBuffer` why does not it return the `baseTransform` also?

kkinnunen-apple replied:
Because this is work towards making CanvasBase not own the ImageBuffer.

shallawa replied:
Please add a FIXME comment saying who should return the `baseTransform` if `CanvasBase` does not own the `ImageBuffer`

shallawa commented on Source/WebCore/html/canvas/PaintRenderingContext2D.cpp:
```diff
@@ -26,22 +26,64 @@
+    if (!m_recordingContext)
+        return;
+    auto& displayList = m_recordingContext->displayList();
+    if (!displayList.isEmpty()) {
```
Early return?

shallawa commented on Source/WebCore/html/canvas/PaintRenderingContext2D.cpp:
```diff
@@ -26,22 +26,64 @@
 
+CustomPaintCanvas& PaintRenderingContext2D::canvas() const
+{
+    return downcast<CustomPaintCanvas>(canvasBase());
```
So `PaintRenderingContext2D` will never use the ImageBuffer which is owned by CanvasBase which is the base class of `CustomPaintCanvas`. right?

kkinnunen-apple replied:
Correct.

kkinnunen-apple committed:
CanvasBase contains 2D context specific drawing methods
https://bugs.webkit.org/show_bug.cgi?id=274750
rdar://128788730

Reviewed by Said Abou-Hallawa.

Move drawingContext(), existingDrawingContext(), baseTransform() to
CanvasRenderingContext2DBase as they are only related to 2D context
implementation.

This is work towards CanvasRenderingContext subclasses themselves managing
the copied buffers.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::drawingContext const): Deleted.
(WebCore::CanvasBase::existingDrawingContext const): Deleted.
(WebCore::CanvasBase::baseTransform const): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::hasCreatedImageBuffer const):
* Source/WebCore/html/CustomPaintCanvas.cpp:
(WebCore::CustomPaintCanvas::getContext):
(WebCore::CustomPaintCanvas::replayDisplayList):
(WebCore::CustomPaintCanvas::copiedImage const):
(WebCore::CustomPaintCanvas::baseTransform const): Deleted.
(WebCore::CustomPaintCanvas::drawingContext const): Deleted.
(WebCore::CustomPaintCanvas::existingDrawingContext const): Deleted.
(WebCore::CustomPaintCanvas::replayDisplayListImpl const): Deleted.
* Source/WebCore/html/CustomPaintCanvas.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::didDraw):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::unwindStateStack):
(WebCore::CanvasRenderingContext2DBase::isAccelerated const):
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::clearCanvas):
(WebCore::CanvasRenderingContext2DBase::transformAreaToDevice const):
(WebCore::CanvasRenderingContext2DBase::calculateCompositingBufferRect):
(WebCore::CanvasRenderingContext2DBase::compositeBuffer):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::drawingContext const):
(WebCore::CanvasRenderingContext2DBase::existingDrawingContext const):
(WebCore::CanvasRenderingContext2DBase::baseTransform const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/PaintRenderingContext2D.cpp:
(WebCore::PaintRenderingContext2D::create):
(WebCore::PaintRenderingContext2D::PaintRenderingContext2D):
(WebCore::PaintRenderingContext2D::canvas const):
(WebCore::PaintRenderingContext2D::drawingContext const):
(WebCore::PaintRenderingContext2D::existingDrawingContext const):
(WebCore::PaintRenderingContext2D::baseTransform const):
(WebCore::PaintRenderingContext2D::replayDisplayList const):
* Source/WebCore/html/canvas/PaintRenderingContext2D.h:

Canonical link: https://commits.webkit.org/279912@main
